### DISCUSSION TITLE
Date picker for forms

### DISCUSSION THREAD

**Discussion by josephclaytonhansen** [espanso/espanso#1593] [Feature requests and Ideas]
It would be great to have an HTML style calendar date picker for FormFields: 
![2023-04-24 16_08_44-html date picker - Google Search](https://user-images.githubusercontent.com/45859835/234117498-2f940e58-300e-41de-a783-0872e7dd98a1.png)

**Comment by nhasian** [comment]
+1
I came here to say this as well.

**Comment by ldmpub** [comment]
Hi, is there any news on this useful feature request? (I should confess I'm a little lost in the current status 😥)

**Comment by smeech** [comment]
It's not on the [Roadmap](https://github.com/orgs/espanso/projects/1) and is unlikely to be a high priority, I'm afraid.

It's reasonably easily [scripted](https://gist.github.com/smeech/5175ddb4790f3ee9e61752162155c086#file-popup-yml), however.

**Reply by ldmpub** [reply]
Thanks, for your reply ❤.

I'm new to Espanso and (a little less) new to Python.

I managed to "POC" the behavior I want to pick a date using Shiny ([see and run my code here](https://shinylive.io/py/examples/#code=NobwRAdghgtgpmAXGKAHVA6VBPMAaMAYwHsIAXOcpMAMwCdiYACAZwAsBLCbJjmVYnTJMAgujxMAkhFQBXMiwkB5eXIUSAynBYsOpCXThRCZDgDc4E2RwA6EOwGImAEQ4m9EaB0NMBsukxkdFAAJtY+ADbaTABWxP4sTJRM9FAQAOdQHCx2cQkA+vRMALxMIHZMlUwA5ACypCFQ2NWINRGyECEc1XgVVdUAKrLajc2t1TBQdF09fZXVAOpwIRAjTS018HSEhjO9EFU1A2wJoxvVMXCye3M1AGJ0HGfjFp273fuH1RpQZP7PNRYsGWH1u3w6AOqXUmEEIbDgswgAF87HY0Kh8tYSkxrFgoABzOCFdpPAAUt1xXDU+UaFHJYA0nG4zl+cAACm4ANZwOiSEIARhs+Fuh1FoqFLIorHibCyZAAl3AhZ8xarDqMWMQaAB3OBwTldIEAIyiIWKwAAbBIAAwAXRVarVETS+NkBLgxSF9GVIsdh11+pYZCmZGK-IdftFNEEk1DQpCIQA9DAYInsOnsD6DpGxdqnmQ2J6wPz+dbUAAPIW+yMASgjlJk8nyxlMpHyRvkZFIpKeRbg5Y4ZGVTGdRrgESLIjyFACzqYtKVYBrdmX9k6cBorB5FjoPcbZFa0jUiiY8TIalaKnP8hPLG0ulIrS0Og8NcQtwAAoYW+Y4BgNzQcAmKi2aVF+RjuBY-6vGQe5qP+A5kKuhxQCw2CwvOG5MPkpJvtWTBOAMwRhN4cAjmReQBMkqQZFkOSgaKd5RCYyw0qy2JUvIGCMlw2CSuyXI8ny-K4fhlH5MkpRMUBFAhGxFAYEGdA0KY8CktUACkIjVMhqriUUpSUSwhR0BghKweJlASJZEC6WKqCPOQpI0BK7EsPKzGtp4ipMK0ID6XQSJlNJLFyQuilBCpfBwOpGlJhpqYaQAmjpQVCnZopQNqcpbi+pAYIQETEHeokMWi6DYmIqCkuimIcBId50DuNZgEitpAA)).

What's the best way to "port" this behavior to a desktop script launched from Espanso ?

**Reply by smeech** [reply]
> I managed to "POC" the behavior I want to pick a date using Shiny ([see and run my code here](https://shinylive.io/py/examples/#code=NobwRAdghgtgpmAXGKAHVA6VBPMAaMAYwHsIAXOcpMAMwCdiYACAZwAsBLCbJjmVYnTJMAgujxMAkhFQBXMiwkB5eXIUSAynBYsOpCXThRCZDgDc4E2RwA6EOwGImAEQ4m9EaB0NMBsukxkdFAAJtY+ADbaTABWxP4sTJRM9FAQAOdQHCx2cQkA+vRMALxMIHZMlUwA5ACypCFQ2NWINRGyECEc1XgVVdUAKrLajc2t1TBQdF09fZXVAOpwIRAjTS018HSEhjO9EFU1A2wJoxvVMXCye3M1AGJ0HGfjFp273fuH1RpQZP7PNRYsGWH1u3w6AOqXUmEEIbDgswgAF87HY0Kh8tYSkxrFgoABzOCFdpPAAUt1xXDU+UaFHJYA0nG4zl+cAACm4ANZwOiSEIARhs+Fuh1FoqFLIorHibCyZAAl3AhZ8xarDqMWMQaAB3OBwTldIEAIyiIWKwAAbBIAAwAXRVarVETS+NkBLgxSF9GVIsdh11+pYZCmZGK-IdftFNEEk1DQpCIQA9DAYInsOnsD6DpGxdqnmQ2J6wPz+dbUAAPIW+yMASgjlJk8nyxlMpHyRvkZFIpKeRbg5Y4ZGVTGdRrgESLIjyFACzqYtKVYBrdmX9k6cBorB5FjoPcbZFa0jUiiY8TIalaKnP8hPLG0ulIrS0Og8NcQtwAAoYW+Y4BgNzQcAmKi2aVF+RjuBY-6vGQe5qP+A5kKuhxQCw2CwvOG5MPkpJvtWTBOAMwRhN4cAjmReQBMkqQZFkOSgaKd5RCYyw0qy2JUvIGCMlw2CSuyXI8ny-K4fhlH5MkpRMUBFAhGxFAYEGdA0KY8CktUACkIjVMhqriUUpSUSwhR0BghKweJlASJZEC6WKqCPOQpI0BK7EsPKzGtp4ipMK0ID6XQSJlNJLFyQuilBCpfBwOpGlJhpqYaQAmjpQVCnZopQNqcpbi+pAYIQETEHeokMWi6DYmIqCkuimIcBId50DuNZgEitpAA)).

I've not been able to get your code to run, unfortunately, or even a minimal `shiny` server that responds to a button click (with help from ChatGPT), I'm afraid.

During the process, and with what little success I had, it became apparent that this is a very slow way of returning a selected date, with the need to start a web-server, open a browser window, and close them both afterwards. The script interpreter startup times I noted on [Reddit](https://www.reddit.com/r/espanso/comments/1lo5rfx/interpreter_startup_times_compared/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) mean that I wouldn't tolerate PowerShell's overhead, never mind an expansion that takes several seconds to complete.

If you're using Linux, then:
```yml
  - trigger: :pickdate
    replace: "{{selected_date}}"
    vars:
      - name: selected_date
        type: shell
        params:
          cmd: zenity --calendar --date-format="%A %d/%m/%Y"
```
is much simpler. I am sure we could come up with a PowerShell equivalent for Windows.

**Reply by ldmpub** [reply]
Thanks a lot for your try.
I agree, even if it works, my solution, based on Shiny, will have very low performances ... 

My OS is Windows 10 and  I runned your command (1) under WSL => it works perfectly ! (it's a very clever script  👍)
Now I need to find a way to call this Linux script externaly (ie "from Windows" and then from Espanso).

(1)`zenity --calendar --date-format="%A %d/%m/%Y"`

**Reply by smeech** [reply]
> Now I need to find a way to call this Linux script externaly (ie "from Windows" and then from Espanso).

It should just be a matter of adding:
```yml
          shell: wsl
```
to the end of the script above, as described [here](https://espanso.org/docs/matches/extensions/#choosing-the-shell), if I've understood your question correctly.

**Comment by ldmpub** [comment]
Thanks to @smeech 👍👍👍 I've finally managed to set up an efficient (but not perfect, see below) "short code" solution using **wsl and YAD** to auto-type a date after day selection in a nice datepicker / calendar.

My Espanso match is as follow:

```yml
  - trigger: "cal "
    replace: "{{Calendrier}} "
    vars:
      - name: Calendrier
        type: shell
        params:
          cmd: "wsl LANG=fr_FR.UTF-8 yad --calendar --show-weeks --date-format=\"%A %d/%m/%Y\""

```

**Steps done (in short)**

1. install wsl (I'm running Windows)
2. in Ubuntu
2.1 install YAD
2.2 install french locale
3. create espanso match (see above)


Note: I have also tried an other GUI calendar (in addition to the suggested zenity and finally selected yad): kdialog.

**There is place for improvments:**
1. hitting the "Cancel" button generates an Espanso error
2. the calendar does not always get focus and starts "hidden" in Windows task bar 🤨
3. the position of the calendar is random (even when using --posx --posy --geometry or other parameters for YAD)

**Reply by smeech** [reply]
ChatGPT and I came up with a PowerShell script which I've  added to the aforementioned [Gist](https://gist.github.com/smeech/5175ddb4790f3ee9e61752162155c086#file-popup_pwsh-yml). 
It works in my Windows10 VM.

**Reply by ldmpub** [reply]
@smeech: I confirm, your script "popup_pwsh.yml" is also working for me (with the same pb mine has using YAD => the calendar dialog is not always displayed in the foreground).

**Comment by ldmpub** [comment]
Little improvement => add `--escape-ok` argument to quicklly close DatePicker and return Today when pressing "Escape" on keyboard.
